fix: ensure validation is run on correct files

beckermr opened the pull request:
This PR ensures we run output validation only on the packages we built.

closes #342 
closes #341

beckermr commented:
cc @isuruf

isuruf commented on recipe/conda_forge_ci_setup/feedstock_outputs.py:
```diff
@@ -140,7 +141,13 @@ def main(feedstock_name):
 
     if is_conda_forge_output_validation_on():
-        distributions = [os.path.relpath(p, conda_build.config.croot) for p in built_distributions()]
+        allowed_dist_names, allowed_subdirs = get_built_distribution_names_and_subdirs()
```
This is an expensive operation. Can we cache this on disk using some python library.

beckermr replied:
Yes let me check into that.

isuruf replied:
example: https://stackoverflow.com/a/23526717/4768820

beckermr replied:
yes this one too: https://pypi.org/project/cache-to-disk/

isuruf commented on recipe/conda_forge_ci_setup/utils.py:
```diff
@@ -9,9 +10,68 @@
+        "FEEDSTOCK_ROOT",
+        os.getcwd(),
+    )
+    if recipe_dir is None:
+        recipe_dir = os.path.join(feedstock_root, "recipe")
+    if build_tool is None:
+        build_tool = determine_build_tool(feedstock_root)
+    if variant is None:
+        if "CONFIG_FILE" in os.environ:
+            variant = [os.environ.get("CONFIG_FILE")]
+        else:
+            variant = [
+                os.path.join(
+                    os.environ.get("CI_SUPPORT", os.path.join(feedstock_root, ".ci_support")),
+                    os.environ.get("CONFIG") + ".yaml"
+                )
+            ]
```
This is fine to have, but we should deprecate these.

beckermr replied:
Yeah it'd be nice to cleanup the set of supplied env vars. It differs by platform and possibly provider in smithy. 

Which ones did you want to deprecate?

isuruf replied:
`recipe_dir` being None and `variant` being None. They should be passed through to this function.

isuruf replied:
Yes, I understand why there needs to be default values for now. We should fix smithy, so that it is not needed in the future.

isuruf commented on recipe/conda_forge_ci_setup/inspect_artifacts.py:
```diff
@@ -1,15 +1,29 @@
 
 
 @click.command()
 def main():
```
for eg: `variant` and `recipe_dir` should be added as `click` options and passed through

beckermr replied:
xref: https://github.com/conda-forge/conda-forge-ci-setup-feedstock/issues/345

beckermr commented:
ok @isuruf I've added caching and I added the options to the CI setup calls. They are not required right now. Once we adjust smithy, we can turn them to be required if we like.

jaimergp commented:
I think this broke `inspect_artifacts` in `staged-recipes`. Not sure why we need to render the recipes there (again). [See here for Linux logs (fails but error is ignored)](https://dev.azure.com/conda-forge/feedstock-builds/_build/results?buildId=1031239&view=logs&jobId=6f142865-96c3-535c-b7ea-873d86b887bd&j=6f142865-96c3-535c-b7ea-873d86b887bd&t=22b0682d-ab9e-55d7-9c79-49f3c3ba4823):

```
Inspecting artifacts
+ command -v inspect_artifacts
+ inspect_artifacts
Error: non-existent: /home/conda/staged-recipes/recipe
+ echo 'inspect_artifacts needs conda-forge-ci-setup >=4.6.0'
inspect_artifacts needs conda-forge-ci-setup >=4.6.0
```

The `Error: non-existent` string comes from `conda_build.render`. I guess in staged-recipes we can't pass `/recipe` blindly, since the recipe dir depends on the PR contents.

beckermr commented:
Ah I see. The goal in this pr was to not print inspection for every artifact but only the ones that were new. When there are downstream tests, we inspect the downstreams too which is odd and confusing.

jaimergp commented:
Got it. I think we raaaarely use downstreams in staged-recipes, right? Should we special case it? Or adjust this so it accepts several directories and pass `recipes/*` from staged-recipes?
